flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicFilteringRuleTest.java

```diff
@@ -0,0 +1,276 @@
+import org.apache.flink.table.planner.factories.TestValuesCatalog;
+import org.apache.flink.table.planner.utils.BatchTableTestUtil;
+import org.apache.flink.table.planner.utils.TableTestBase;
+
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Test for rules that extend {@link DynamicFilteringRule} to create {@link
+ * org.apache.flink.table.planner.plan.nodes.physical.batch.BatchPhysicalDynamicFilteringTableSourceScan}.
+ */
+public class DynamicFilteringRuleTest extends TableTestBase {
+    protected BatchTableTestUtil util = batchTestUtil(TableConfig.getDefault());
+    TestValuesCatalog catalog = new TestValuesCatalog("testCatalog", "test_database", true);
+
+    @Before
+    public void setup() {
+        catalog.open();
+        util.tableEnv().registerCatalog("testCatalog", catalog);
+        util.tableEnv().useCatalog("testCatalog");
+        TableConfig tableConfig = util.tableEnv().getConfig();
+        tableConfig.set(OptimizerConfigOptions.TABLE_OPTIMIZER_DYNAMIC_FILTERING_ENABLED, true);
+
+        // partition fact table.
+        String ddl3 =
+                "CREATE TABLE test_database.fact_part (\n"
+                        + "  id BIGINT,\n"
+                        + "  name STRING,\n"
+                        + "  amount BIGINT,\n"
+                        + "  price BIGINT,\n"
+                        + "  fact_date_sk BIGINT\n"
+                        + ") PARTITIONED BY (fact_date_sk)\n"
+                        + "WITH (\n"
+                        + " 'connector' = 'values',\n"
+                        + " 'partition-list' = 'fact_date_sk:1990;fact_date_sk:1991;fact_date_sk:1992',\n"
+                        + " 'dynamic-filtering-fields' = 'fact_date_sk;amount',\n"
+                        + " 'bounded' = 'true'\n"
+                        + ")";
+        util.tableEnv().executeSql(ddl3);
+
+        // dim table.
+        String ddl2 =
+                "CREATE TABLE test_database.dim (\n"
+                        + "  id BIGINT,\n"
+                        + "  name STRING,\n"
+                        + "  amount BIGINT,\n"
+                        + "  price BIGINT,\n"
+                        + "  dim_date_sk BIGINT\n"
+                        + ") WITH (\n"
+                        + " 'connector' = 'values',\n"
+                        + " 'bounded' = 'true'\n"
+                        + ")";
+        util.tableEnv().executeSql(ddl2);
+    }
+
+    @Test
+    public void testDimTableFilteringFieldsNotInJoinKey() {
+        // fact_part.id not in dynamic-filtering-fields, so dynamic filtering will not succeed.
+        String query =
+                "Select * from dim, fact_part where fact_part.id = dim.id and dim.price < 500";
+        util.verifyRelPlan(query);
+    }
+
+    @Test
+    public void testDynamicFilteringFactInRightRule() throws TableNotExistException {
+        CatalogTableStatistics tableStatistics = new CatalogTableStatistics(1, 1, 1, 1);
```

Comment:
why we need statistics

Reply:
> why we need statistics

Adding statistics to add a exchange node upper table source scan.